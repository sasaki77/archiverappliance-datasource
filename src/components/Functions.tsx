import React from 'react';
import { InlineFieldRow, InlineLabel } from '@grafana/ui';
import { createFuncDescriptor } from '../aafunc';
import { FuncDef, FunctionDescriptor } from '../types';

import { FunctionElem } from './FunctionElem';
import { FunctionAdd } from './FunctionAdd';

export interface FunctionsProps {
  funcs: FunctionDescriptor[];
  onChange: (func: FunctionDescriptor[]) => void;
  onRunQuery: () => void;
}

const arrayMove = (array: any[], from: number, to: number): any[] => {
  const newArray = [...array];
  const moveElem = newArray.splice(from, 1)[0];
  newArray.splice(to, 0, moveElem);

  return newArray;
};

class Functions extends React.PureComponent<FunctionsProps> {
  constructor(props: FunctionsProps) {
    super(props);
  }

  addFunction = (funcDef: FuncDef) => {
    const { onChange, onRunQuery } = this.props;
    const newFunc = createFuncDescriptor(funcDef, funcDef.defaultParams);
    const newFuncs = [...this.props.funcs];
    newFuncs.push(newFunc);
    onChange(newFuncs);

    if (newFunc.params.length || newFunc.def.params.length === 0) {
      onRunQuery();
    }
  };

  handleRemoveFunction = (func: FunctionDescriptor, index: number) => {
    const { onChange, onRunQuery } = this.props;
    const newFuncs = [...this.props.funcs];
    newFuncs.splice(index, 1);
    onChange(newFuncs);
    onRunQuery();
  };

  handleMoveLeft = (func: FunctionDescriptor, index: number) => {
    if (index < 1) {
      return;
    }

    const { onChange, onRunQuery } = this.props;
    const newFuncs = arrayMove(this.props.funcs, index, index - 1);
    onChange(newFuncs);
    onRunQuery();
  };

  handleMoveRight = (func: FunctionDescriptor, index: number) => {
    if (index >= this.props.funcs.length - 1) {
      return;
    }

    const { onChange, onRunQuery } = this.props;
    const newFuncs = arrayMove(this.props.funcs, index, index + 1);
    onChange(newFuncs);
    onRunQuery();
  };

  onFuncChange = (func: FunctionDescriptor, index: number) => {
    const { onChange } = this.props;
    const newFuncs = [...this.props.funcs];
    newFuncs.splice(index, 1, func);
    onChange(newFuncs);
  };

  render() {
    const { funcs, onRunQuery } = this.props;
    return (
      <InlineFieldRow>
        <InlineLabel
          width={12}
          interactive={true}
          tooltip={
            <p>
              Functions are used to apply post processing to the data. You can add, move and remove functions. Select
              preffered functions from categoires. Some functions require parameters. You can edit parameters after
              adding the function. Functions are applied from left to right.
            </p>
          }
        >
          Functions
        </InlineLabel>
        {funcs &&
          funcs.map((func, index) => (
            <FunctionElem
              key={index}
              index={index}
              func={func}
              onMoveLeft={this.handleMoveLeft}
              onMoveRight={this.handleMoveRight}
              onRemove={this.handleRemoveFunction}
              onChange={this.onFuncChange}
              onRunQuery={onRunQuery}
            />
          ))}
        <FunctionAdd addFunc={this.addFunction} />
        <div className="gf-form-label gf-form-label--grow"></div>
      </InlineFieldRow>
    );
  }
}

export { Functions };
