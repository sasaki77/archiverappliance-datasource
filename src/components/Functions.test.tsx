import React from 'react';
import { render } from '@testing-library/react';

import { Functions, FunctionsProps } from './Functions';

const setup = (propOverrides?: object) => {
  const props: FunctionsProps = {
    funcs: [],
    onChange: jest.fn(),
    onRunQuery: jest.fn(),
  };

  Object.assign(props, propOverrides);

  return render(<Functions {...props} />);
};

describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with functions', () => {
    const funcs = [
      {
        def: {
          category: 'Filter Series',
          defaultParams: [5, 'avg'],
          name: 'top',
          params: [
            {
              name: 'number',
              type: 'int',
            },
            {
              name: 'value',
              options: ['avg', 'min', 'max', 'absoluteMin', 'absoluteMax', 'sum'],
              type: 'string',
            },
          ],
        },
        params: ['5', 'avg'],
        text: 'top(5, avg)',
      },
      {
        def: {
          category: 'Transform',
          defaultParams: [100],
          name: 'scale',
          params: [
            {
              name: 'factor',
              type: 'float',
            },
          ],
        },
        params: ['0.001'],
        text: 'scale(0.001)',
      },
    ];
    const wrapper = setup({ funcs: funcs });
    expect(wrapper).toMatchSnapshot();
  });
});
