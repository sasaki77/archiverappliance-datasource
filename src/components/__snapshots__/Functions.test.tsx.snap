// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="gf-form-inline"
      >
        <div
          class="gf-form"
        >
          <label
            class="query-keyword gf-form-label width-6"
          >
            Functions
            <div
              class="css-wf08df-Icon"
              tabindex="0"
            >
              <svg
                data-testid="info-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </label>
        </div>
        <div
          class="gf-form"
        />
        <div
          class="gf-form"
        >
          <button
            class="css-z53gi5-button"
            type="button"
          >
            <div
              class="css-wf08df-Icon"
            >
              <svg
                data-testid="plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
            <span
              class="css-1mhnkuh"
            >
              Add
              <div
                class="css-wf08df-Icon"
              >
                <svg
                  data-testid="angle-down"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </span>
          </button>
        </div>
        <div
          class="gf-form gf-form--grow"
        >
          <div
            class="gf-form-label gf-form-label--grow"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="gf-form-inline"
    >
      <div
        class="gf-form"
      >
        <label
          class="query-keyword gf-form-label width-6"
        >
          Functions
          <div
            class="css-wf08df-Icon"
            tabindex="0"
          >
            <svg
              data-testid="info-circle"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
        </label>
      </div>
      <div
        class="gf-form"
      />
      <div
        class="gf-form"
      >
        <button
          class="css-z53gi5-button"
          type="button"
        >
          <div
            class="css-wf08df-Icon"
          >
            <svg
              data-testid="plus"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
          <span
            class="css-1mhnkuh"
          >
            Add
            <div
              class="css-wf08df-Icon"
            >
              <svg
                data-testid="angle-down"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </span>
        </button>
      </div>
      <div
        class="gf-form gf-form--grow"
      >
        <div
          class="gf-form-label gf-form-label--grow"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Render should render with functions 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="gf-form-inline"
      >
        <div
          class="gf-form"
        >
          <label
            class="query-keyword gf-form-label width-6"
          >
            Functions
            <div
              class="css-wf08df-Icon"
              tabindex="0"
            >
              <svg
                data-testid="info-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </label>
        </div>
        <div
          class="gf-form"
        >
          <span
            class="gf-form-label query-part"
          >
            <span
              style="cursor: pointer;"
              tabindex="0"
            >
              top
            </span>
            <span>
              (
            </span>
            
            <div>
              <div
                class="aa-func-param"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="react-autowhatever-1"
                  class="react-autosuggest__container"
                  role="combobox"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="react-autowhatever-1"
                    autocomplete="off"
                    class="tight-form-func-param"
                    placeholder="number"
                    spellcheck="false"
                    style="width: 1ch; font-family: Consolas, "Courier New", Courier, Monaco, monospace;"
                    type="text"
                    value="5"
                  />
                  <div
                    class="react-autosuggest__suggestions-container"
                    id="react-autowhatever-1"
                    role="listbox"
                  />
                </div>
              </div>
            </div>
            <span
              class="comma"
            >
              ,Â 
            </span>
            <div>
              <div
                class="aa-func-param"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="react-autowhatever-1"
                  class="react-autosuggest__container"
                  role="combobox"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="react-autowhatever-1"
                    autocomplete="off"
                    class="tight-form-func-param"
                    placeholder="value"
                    spellcheck="false"
                    style="width: 3ch; font-family: Consolas, "Courier New", Courier, Monaco, monospace;"
                    type="text"
                    value="avg"
                  />
                  <div
                    class="react-autosuggest__suggestions-container"
                    id="react-autowhatever-1"
                    role="listbox"
                  />
                </div>
              </div>
            </div>
            <span>
              )
            </span>
          </span>
          <span
            class="gf-form-label query-part"
          >
            <span
              style="cursor: pointer;"
              tabindex="0"
            >
              scale
            </span>
            <span>
              (
            </span>
            
            <div>
              <div
                class="aa-func-param"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="react-autowhatever-1"
                  class="react-autosuggest__container"
                  role="combobox"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="react-autowhatever-1"
                    autocomplete="off"
                    class="tight-form-func-param"
                    placeholder="factor"
                    spellcheck="false"
                    style="width: 5ch; font-family: Consolas, "Courier New", Courier, Monaco, monospace;"
                    type="text"
                    value="0.001"
                  />
                  <div
                    class="react-autosuggest__suggestions-container"
                    id="react-autowhatever-1"
                    role="listbox"
                  />
                </div>
              </div>
            </div>
            <span>
              )
            </span>
          </span>
        </div>
        <div
          class="gf-form"
        >
          <button
            class="css-z53gi5-button"
            type="button"
          >
            <div
              class="css-wf08df-Icon"
            >
              <svg
                data-testid="plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
            <span
              class="css-1mhnkuh"
            >
              Add
              <div
                class="css-wf08df-Icon"
              >
                <svg
                  data-testid="angle-down"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </span>
          </button>
        </div>
        <div
          class="gf-form gf-form--grow"
        >
          <div
            class="gf-form-label gf-form-label--grow"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="gf-form-inline"
    >
      <div
        class="gf-form"
      >
        <label
          class="query-keyword gf-form-label width-6"
        >
          Functions
          <div
            class="css-wf08df-Icon"
            tabindex="0"
          >
            <svg
              data-testid="info-circle"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
        </label>
      </div>
      <div
        class="gf-form"
      >
        <span
          class="gf-form-label query-part"
        >
          <span
            style="cursor: pointer;"
            tabindex="0"
          >
            top
          </span>
          <span>
            (
          </span>
          
          <div>
            <div
              class="aa-func-param"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="react-autowhatever-1"
                class="react-autosuggest__container"
                role="combobox"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="react-autowhatever-1"
                  autocomplete="off"
                  class="tight-form-func-param"
                  placeholder="number"
                  spellcheck="false"
                  style="width: 1ch; font-family: Consolas, "Courier New", Courier, Monaco, monospace;"
                  type="text"
                  value="5"
                />
                <div
                  class="react-autosuggest__suggestions-container"
                  id="react-autowhatever-1"
                  role="listbox"
                />
              </div>
            </div>
          </div>
          <span
            class="comma"
          >
            ,Â 
          </span>
          <div>
            <div
              class="aa-func-param"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="react-autowhatever-1"
                class="react-autosuggest__container"
                role="combobox"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="react-autowhatever-1"
                  autocomplete="off"
                  class="tight-form-func-param"
                  placeholder="value"
                  spellcheck="false"
                  style="width: 3ch; font-family: Consolas, "Courier New", Courier, Monaco, monospace;"
                  type="text"
                  value="avg"
                />
                <div
                  class="react-autosuggest__suggestions-container"
                  id="react-autowhatever-1"
                  role="listbox"
                />
              </div>
            </div>
          </div>
          <span>
            )
          </span>
        </span>
        <span
          class="gf-form-label query-part"
        >
          <span
            style="cursor: pointer;"
            tabindex="0"
          >
            scale
          </span>
          <span>
            (
          </span>
          
          <div>
            <div
              class="aa-func-param"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="react-autowhatever-1"
                class="react-autosuggest__container"
                role="combobox"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="react-autowhatever-1"
                  autocomplete="off"
                  class="tight-form-func-param"
                  placeholder="factor"
                  spellcheck="false"
                  style="width: 5ch; font-family: Consolas, "Courier New", Courier, Monaco, monospace;"
                  type="text"
                  value="0.001"
                />
                <div
                  class="react-autosuggest__suggestions-container"
                  id="react-autowhatever-1"
                  role="listbox"
                />
              </div>
            </div>
          </div>
          <span>
            )
          </span>
        </span>
      </div>
      <div
        class="gf-form"
      >
        <button
          class="css-z53gi5-button"
          type="button"
        >
          <div
            class="css-wf08df-Icon"
          >
            <svg
              data-testid="plus"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
          <span
            class="css-1mhnkuh"
          >
            Add
            <div
              class="css-wf08df-Icon"
            >
              <svg
                data-testid="angle-down"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </span>
        </button>
      </div>
      <div
        class="gf-form gf-form--grow"
      >
        <div
          class="gf-form-label gf-form-label--grow"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
